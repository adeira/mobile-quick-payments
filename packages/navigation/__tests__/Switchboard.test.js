// @flow

import * as React from 'react';
import TestRenderer from 'react-test-renderer';

import { RegisterSwitch, Switchboard, Switch } from '../index';

const MockedScene = ({
  shouldThrow,
  message
}: {|
  +shouldThrow?: boolean,
  +message?: string
|}) => {
  if (shouldThrow) {
    throw new Error('This component should not be rendered.');
  }
  return message || 'ok';
};

it('renders only the first registered component', () => {
  const Application = (
    <Switchboard>
      <RegisterSwitch path="/A" component={<MockedScene />} />
      <RegisterSwitch
        path="/B"
        component={<MockedScene shouldThrow={true} />}
      />
      <RegisterSwitch
        path="/C"
        component={<MockedScene shouldThrow={true} />}
      />
    </Switchboard>
  );

  expect(
    TestRenderer.create(Application).getInstance().state
  ).toMatchSnapshot();

  expect(TestRenderer.create(Application).toJSON()).toBe('ok');
});

it('throw an error when you try to register two switches with the same name', () => {
  expect(() =>
    TestRenderer.create(
      <Switchboard>
        <RegisterSwitch path="/A" component={<MockedScene />} />
        <RegisterSwitch path="/A" component={<MockedScene />} />
      </Switchboard>
    )
  ).toThrow(
    "Route with path '/A' already exists and therefore you cannot register it again."
  );
});

it('switches the scenes declaratively', () => {
  const Application = (
    <Switchboard>
      <RegisterSwitch
        path="/A"
        component={
          <React.Fragment>
            <Switch to="/B" />
            <MockedScene /> {/* This component will be mounted and rendered! */}
          </React.Fragment>
        }
      />
      <RegisterSwitch path="/B" component={<MockedScene message="BBB" />} />
      <RegisterSwitch
        path="/C"
        component={<MockedScene shouldThrow={true} />}
      />
    </Switchboard>
  );

  expect(
    TestRenderer.create(Application).getInstance().state
  ).toMatchSnapshot();

  expect(TestRenderer.create(Application).toJSON()).toBe('BBB');
});

it('throws when you try to switch to non-existent route', () => {
  const Application = (
    <Switchboard>
      <RegisterSwitch
        path="/"
        component={
          <React.Fragment>
            <Switch to="/unknown-route-wtf" />
          </React.Fragment>
        }
      />
    </Switchboard>
  );

  expect(() => TestRenderer.create(Application)).toThrowErrorMatchingSnapshot();
});
