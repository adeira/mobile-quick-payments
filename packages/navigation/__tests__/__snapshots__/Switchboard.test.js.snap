// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders only the first registered component 1`] = `
Object {
  "actions": Object {
    "activateRoute": [Function],
    "isRouteActive": [Function],
    "registerRoute": [Function],
  },
  "routes": Array [
    Object {
      "component": <MockedScene />,
      "isActive": true,
      "path": "/A",
    },
    Object {
      "component": <MockedScene
        shouldThrow={true}
      />,
      "isActive": false,
      "path": "/B",
    },
    Object {
      "component": <MockedScene
        shouldThrow={true}
      />,
      "isActive": false,
      "path": "/C",
    },
  ],
}
`;

exports[`switches the scenes declaratively 1`] = `
Object {
  "actions": Object {
    "activateRoute": [Function],
    "isRouteActive": [Function],
    "registerRoute": [Function],
  },
  "routes": Array [
    Object {
      "component": <React.Fragment>
        <Switch
          to="/B"
        />
        <MockedScene />
         
      </React.Fragment>,
      "isActive": false,
      "path": "/A",
    },
    Object {
      "component": <MockedScene
        message="BBB"
      />,
      "isActive": true,
      "path": "/B",
    },
    Object {
      "component": <MockedScene
        shouldThrow={true}
      />,
      "isActive": false,
      "path": "/C",
    },
  ],
}
`;

exports[`throws when you try to switch to non-existent route 1`] = `"Cannot switch to a path '/unknown-route-wtf' because this path doesn't exist (or component RegisterSwitch is not mounted yet)."`;
